hivemind/utils/mpfuture.py

```diff
@@ -2,171 +2,253 @@
+import os
+import threading
+import uuid
+from enum import Enum, auto
+from typing import Generic, TypeVar, Dict, Optional, Any, Callable
 
-from hivemind.utils.threading import run_in_background
+import torch
 
-ResultType = TypeVar('ResultType')
+from hivemind.utils.logging import get_logger
 
 
-class FutureStateError(RuntimeError):
-    """Raised when attempting to change state of a future in a terminal state (e.g. finished)"""
-    pass
+logger = get_logger(__name__)
 
+# flavour types
+ResultType = TypeVar('ResultType')
+PID, UID, State, PipeEnd = int, int, Any, mp.connection.Connection
+ALL_STATES = base.PENDING, base.RUNNING, base.FINISHED, base.CANCELLED, base.CANCELLED_AND_NOTIFIED
+TERMINAL_STATES = {base.FINISHED, base.CANCELLED, base.CANCELLED_AND_NOTIFIED}
 
-class MPFuture(base.Future, Generic[ResultType]):
-    """ Multiprocessing version of concurrent.futures.Future. Can also be awaited like asyncio.Future """
+try:
+    from concurrent.futures import InvalidStateError
+except ImportError:
+    # Python 3.7 doesn't raise concurrent.futures.InvalidStateError for repeating set_result/set_exception calls and
+    # doesn't even define this error. In this module, we simulate the Python 3.8+ behavior,
+    # defining and raising this error if necessary.
+    class InvalidStateError(Exception):
+        """Raised when attempting to change state of a future in a terminal state (e.g. finished)"""
 
-    TERMINAL_STATES = {base.FINISHED, base.CANCELLED, base.CANCELLED_AND_NOTIFIED}
 
-    def __init__(self, connection: mp.connection.Connection):
-        """ manually create MPFuture. Please use MPFuture.make_pair instead """
+class UpdateType(Enum):
+    RESULT = auto()
+    EXCEPTION = auto()
+    CANCEL = auto()
+
+
+class MPFuture(base.Future, Generic[ResultType]):
+    """
+    A version of concurrent.futures.Future / asyncio.Future that can be fulfilled from a separate process.
+    Any process can access future status and set the result / exception and check for state.
+    However, only the original process (i.e. the process that created the future) can await the result or exception.
+
+    :note: This is an internal primitive that is not guaranteed to work outside of hivemind applications.
+     More specifically, there are two known limitations:
+       - MPFuture works between processes created through inheritance (e.g. fork), *not* for independent processes
+       - Different executors (non-origin processes) cannot call set_result / set_exception / cancel simultaneously
+    """
+    lock = mp.Lock()  # global lock that prevents simultaneous initialization and writing
+    pipe_waiter_thread: Optional[threading.Thread] = None  # process-specific thread that receives results/exceptions
+    global_mpfuture_receivers: Dict[PID, PipeEnd] = mp.Manager().dict()
+    active_futures: Optional[Dict[PID, MPFuture]] = None  # pending or running futures originated from current process
```

Comment:
```suggestion
    active_futures: Optional[Dict[UID, MPFuture]] = None  # pending or running futures originated from current process
```

Reply:
I suggest to protect all interactions with `active_futures` with a `threading.Lock`. Otherwise, it seems to be not thread-safe (e.g. if several processes use even different `MPFuture` instances).